src/operator/optimizer_op-inl.h

```diff
@@ -485,6 +485,95 @@ inline void SGDMomUpdateEx(const nnvm::NodeAttrs& attrs,
   }
 }
 
+
+struct FTMLParam : public dmlc::Parameter<FTMLParam> {
+  float lr;
+  float beta1;
+  float beta2;
+  double epsilon;
+  int t;
+  float wd;
+  float rescale_grad;
+  float clip_grad;
+  DMLC_DECLARE_PARAMETER(FTMLParam) {
+    DMLC_DECLARE_FIELD(lr)
+    .describe("Learning rate.");
+    DMLC_DECLARE_FIELD(beta1)
+    .set_default(0.6f)
+    .set_range(0.0f, 1.0f)
+    .describe("Generally close to 0.5.");
+    DMLC_DECLARE_FIELD(beta2)
+    .set_default(0.999f)
+    .set_range(0.0f, 1.0f)
+    .describe("Generally close to 1.");
+    DMLC_DECLARE_FIELD(epsilon)
+    .set_default(1e-8f)
+    .describe("Epsilon to prevent div 0.");
+    DMLC_DECLARE_FIELD(t)
+    .describe("Number of update.");
+    DMLC_DECLARE_FIELD(wd)
+    .set_default(0.0f)
+    .describe("Weight decay augments the objective function with a "
+              "regularization term that penalizes large weights. "
+              "The penalty scales with the square of the magnitude of each weight.");
+    DMLC_DECLARE_FIELD(rescale_grad)
+    .set_default(1.0f)
+    .describe("Rescale gradient to grad = rescale_grad*grad.");
+    DMLC_DECLARE_FIELD(clip_grad)
+    .set_default(-1.0f)
+    .describe("Clip gradient to the range of [-clip_gradient, clip_gradient] "
+              "If clip_gradient <= 0, gradient clipping is turned off. "
+              "grad = max(min(grad, clip_gradient), -clip_gradient).");
+  }
+};
+
+
+struct FTMLKernel {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, DType* weight, DType* grad,
+    DType* d, DType* v, DType* z, const DType lr, const DType beta1,
+    const DType beta2, const DType epsilon, const DType t,
+    const DType wd, const DType rescale_grad, const DType clip_grad,
+    const OpReqType req) {
+    using namespace mshadow_op;
+    const DType grad_i = clip_grad >= 0.0f
+        ? (clip::Map(rescale_grad * grad[i], clip_grad) + wd * weight[i])
+        : (rescale_grad * grad[i] + wd * weight[i]);
```

Comment:
We should clip after adding the gradient of L2. This is consistent with other optimizers.